### ISSUE TITLE
How install on Amazon Lambda?

### ISSUE THREAD

**Issue by renatoargh** [bbcrd/audiowaveform#35]
[I have successfully installed `audiowaveform` on Amazon Linux](https://github.com/bbcrd/audiowaveform/issues/27) (which is the same OS for Amazon Lambda), but I can't seem to find a way to create a unique executable file that will include all shared libraries in which `audiowaveform` depends upon (running `ldd audiowaveform` shows a lot of them). I can't think of any other way to achieve this. Any hints?

Being able to do this means that I would be able to run it under demand not worrying about provisioning EC2 servers.

**Comment by oncletom** [comment]
Probably best way would be to make a Node.js native addon to wrap the sources and to compile it for the adequate platform target.

**Comment by renatoargh** [comment]
I will try implementing this (node native add-ons) and will post any progress or further questions here.

**Comment by oncletom** [comment]
Perfect @renatoargh :-)

**Comment by jj227** [comment]
hey @renatoargh, did you have any luck?

**Comment by bradddd** [comment]
Has anyone gotten this to work yet?

**Comment by chrisn** [comment]
The article [Running Arbitrary Executables in AWS Lambda](https://aws.amazon.com/blogs/compute/running-executables-in-aws-lambda/) has some useful information:

> If you compile your own binaries, ensure that they’re either statically linked or built for the matching version of Amazon Linux. The current version of Amazon Linux in use within AWS Lambda can always be found on the Supported Versions page of the Lambda docs.

So, the first thing to do is modify [CMakeLists.txt](https://github.com/bbcrd/audiowaveform/blob/master/CMakeLists.txt) to enable (optional) static linking.

**Comment by holmesal** [comment]
Has anyone succeeded in getting this to work?

**Comment by alexwhittemore** [comment]
Hey y'all, in particular @chrisn - I've dug in with CMakeLists.txt and discerned that I need to be linking the .a static copies of the dependencies by setting 
`set(CMAKE_FIND_LIBRARY_SUFFIXES ".a")`
 before all the references to `find_package()` that I want linked statically, for example, before
```
find_package(LibId3Tag REQUIRED)
if(LIBID3TAG_FOUND)
    message(STATUS "LIBID3TAG_INCLUDE_DIR='${LIBID3TAG_INCLUDE_DIR}'")
    message(STATUS "LIBID3TAG_LIBRARY=${LIBID3TAG_LIBRARY}")
    include_directories(${LIBID3TAG_INCLUDE_DIR})
endif(LIBID3TAG_FOUND)
```
I figured out that if I use readelf, I can determine what libraries THESE depend on, for example, 
`readelf -d /usr/lib/libid3tag.so`
gives me
```
Dynamic section at offset 0x17da0 contains 26 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000e (SONAME)             Library soname: [libid3tag.so.0]
```
In other words, libid3tag.so depends on libz and libc (the standard library). If I do this same thing for libz, I find that it depends on only libc, so i've hit the end of the dependency tree.

In order to satisfy this dependency before static linking, I got to the point of including it in my CMakeLists.txt before libid3tag:

```
find_package( ZLIB REQUIRED )
if ( ZLIB_FOUND )
    message(STATUS "ZLIB_INCLUDE_DIRS=${ZLIB_INCLUDE_DIRS}")
    include_directories( ${ZLIB_INCLUDE_DIRS} )
endif( ZLIB_FOUND )
```
In some cases, these dependencies are satisfied by "FindX.cmake" definitions that ship with cmake. In this case, I had to dig around online to find a "Findzlib.cmake" that works to drop in my cmake/modules directory with the others.

So two questions:

First) If I want to statically link this, do I seriously have to chase down the entire dependency tree and add directives to find and include all those things? Is there no shorthand way that cmake does this automatically for me? It feels like a tremendous repetition of labor that someone must have come across before.

Second) This strategy worked great for libid3tag - I found the one dependency, zlib, and included libz.a with the above find_package. Using readelf on the primary target output (audiowaveform), I determine that, yes, libid3tag gets statically linked and no longer shows up as a dynamic dependency. Cool! BUT, I extrapolated the process to LibGD, which has a longer list of deps:
```

# LibGD Requires:
# libjpeg
# libz
# libpng12
# libfreetype
# libfontconfig
# libXpm
# libvpx
# libm
# libtiff
```
Despite putting libz before both LibGD and LibId3Tag, my compliation now fails: 
```
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libpng.a(libpng12_la-png.o): In function `png_reset_crc':
(.text+0x1ab): undefined reference to `crc32'
```
That's super confusing to me, because crc32 is a function in zlib, and should already be statically linked in my output. So why is the linker complaining that it's not found?

I'd sure love a little more guidance from someone who perhaps knows how the toolchain works.

**Comment by georgebohnisch** [comment]
Need this for Lambda as well.

**Comment by alexwhittemore** [comment]
@georgebohnisch Still WIP. I took a break on it for a few days to come back with fresh eyes.

**Comment by holmesal** [comment]
We weren't able to get this working using the method @alexwhittemore was chasing down above, but we did come across a commercial product called [Ermine](http://www.magicermine.com/) that allowed us to bundle this up and deploy it on AWS Lambda - it was quite easy once we bought the license.

It's a paid product so it might not be the right option for everyone here, but I wanted to make it known in case anyone else is in the same boat as us.

**Comment by ffxsam** [comment]
I got this working on Lambda. To save everyone else the heavy lifting, feel free to grab this:

http://ffxsam.s3.amazonaws.com/public/audiowaveform-lambda.zip

When you package up your Lambda function, leave the `bin` and `lib` folders intact. Lambda's environment will automatically have `/var/task/lib` in `LD_LIBRARY_PATH` so no extra work there is needed. However, you will have to add the `bin` folder to the path inside your `exports.handler` function:

```js
process.env.PATH =
    process.env.PATH + ':' + process.env.LAMBDA_TASK_ROOT + '/bin';
```

Then you can use Node.js to execute `audiowaveform`.

I know it's not an ideal solution, so if anyone ever figures out how to statically link all those library files into a single binary, please reply to this thread!

**Comment by chrisn** [comment]
Thanks so much for your contribution, Sam.

**Comment by georgebohnisch** [comment]
Thanks @ffxsam!

**Comment by alexwhittemore** [comment]
If it fits it ships! Great contribution @ffxsam

**Comment by georgebohnisch** [comment]
@ffxsam can you post a sample of your lambda script to get this working?

**Comment by ffxsam** [comment]
@georgebohnisch The code is proprietary so I can't share the whole thing, but this is the relevant part:

```js
execute(`audiowaveform -i ${wavFilename} --pixels-per-second ${pps} -b 8 -o ${peaksFilename}`).then(() => {
  // do something else
});
```

With `execute` defined as:

```js
function execute(command) {
  return new Promise((resolve, reject) => {
    const exec = require('child_process').exec;
    let output = '';
    const child = exec(command, error => {
      if (error) reject(error);
      else resolve(output);
    });

    child.stderr.on('data', data => {
      output += data.toString();
    });
  });
}
```

**Comment by ojczeo** [comment]
@ffxsam how do format  input/output params  for `audiowaveform` (in your example `wavFilename` and `peaksFilename `)?

**Comment by ojczeo** [comment]
I finally resolved problem with Lambda, here's a script, might be useful for somebody:
https://gist.github.com/ojczeo/1f7325b438945bb4878941fb2ebb8409

**Comment by chrisn** [comment]
The `static-link` branch allows audiowaveform to be linked statically against libgd, libsndfile, etc. See README.md for details on how to build. I haven't tried deploying the resulting binary to Lambda, feedback welcome.

**Comment by ajbarber** [comment]
> The `static-link` branch allows audiowaveform to be linked statically against libgd, libsndfile, etc. See README.md for details on how to build. I haven't tried deploying the resulting binary to Lambda, feedback welcome.

I can build statically and deploy to aws lambda successfully, following the instructions on the main README. There is no need to use the archived binaries attached above anymore....

**Comment by ffxsam** [comment]
@ajbarber Great news, and good timing for me! I'll see if I can do a static build on an Amazon Linux EC2 (so it's Lambda-compatible).

**Comment by ffxsam** [comment]
> The `static-link` branch allows audiowaveform to be linked statically against libgd, libsndfile, etc. See README.md for details on how to build. I haven't tried deploying the resulting binary to Lambda, feedback welcome.

@chrisn What is this `static-link` branch? I don't see that branch in this repo.

**Comment by ffxsam** [comment]
@ajbarber I don't know how you got this to build. I tried to install the dependencies:

    $ sudo yum install git make cmake gcc-c++ libmad-devel   libid3tag-devel libsndfile-devel gd-devel boost-devel

and get:

```
No package libmad-devel available.
No package libid3tag-devel available.
No package libsndfile-devel available.
```

**Comment by alexwhittemore** [comment]
@ffxsam looks like it's not actually a branch, but just a makefile flag option that's been added to the current master (haven't tested though).

**Comment by ffxsam** [comment]
Also:

```
-- Build type not specified: default is Release
-- CMAKE_VERSION=2.8.12.2
-- Build type: Release
-- CMAKE_MODULE_PATH='/home/ec2-user/audiowaveform/cmake/modules'
-- Building version 1.3.3
-- Static build
-- Could NOT find ZLIB (missing:  ZLIB_LIBRARY) (found version "1.2.8")
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find PNG (missing: PNG_LIBRARY PNG_PNG_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake/Modules/FindPNG.cmake:105 (find_package_handle_standard_args)
  cmake/modules/FindLibGD.cmake:39 (find_package)
  CMakeLists.txt:81 (find_package)


-- Configuring incomplete, errors occurred!
```

**Comment by ajbarber** [comment]
> Also:
> 
> ```
> -- Build type not specified: default is Release
> -- CMAKE_VERSION=2.8.12.2
> -- Build type: Release
> -- CMAKE_MODULE_PATH='/home/ec2-user/audiowaveform/cmake/modules'
> -- Building version 1.3.3
> -- Static build
> -- Could NOT find ZLIB (missing:  ZLIB_LIBRARY) (found version "1.2.8")
> CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
>   Could NOT find PNG (missing: PNG_LIBRARY PNG_PNG_INCLUDE_DIR)
> Call Stack (most recent call first):
>   /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
>   /usr/share/cmake/Modules/FindPNG.cmake:105 (find_package_handle_standard_args)
>   cmake/modules/FindLibGD.cmake:39 (find_package)
>   CMakeLists.txt:81 (find_package)
> 
> 
> -- Configuring incomplete, errors occurred!
> ```
> Can anyone help with this? These libraries are installed globally. Not sure why cmake is not seeing them. Once I get it solved, I can update the repo's README for instructions specific to Amazon Linux.

Cmake is complaining as the static versions of these libraries (i.e suffix .a) are either not installed on your system (likely) or it can't find them (unlikely) :smile:

**Comment by ajbarber** [comment]
To help anyone who just wants a binary here is my Dockerfile:
~~~~
FROM ubuntu:14.04
RUN apt-get update
RUN apt-get --assume-yes install wget libstdc++6-4.7-dev libc6-dev git make cmake gcc g++ libmad0-dev \
  libid3tag0-dev libsndfile1-dev libgd-dev libboost-filesystem-dev \
  libboost-program-options-dev \
  libboost-regex-dev

COPY ./build.sh build.sh
RUN chmod +x build.sh
RUN ./build.sh
~~~~

and here is the build script `build.sh`

~~~~
#!/bin/bash
git clone https://github.com/bbc/audiowaveform.git
cd audiowaveform
wget https://github.com/google/googletest/archive/release-1.8.0.tar.gz
tar xzf release-1.8.0.tar.gz
ln -s googletest-release-1.8.0/googletest googletest
ln -s googletest-release-1.8.0/googlemock googlemock
mkdir build
cd build
cmake -D BUILD_STATIC=1 ..
make
~~~~

**Comment by ffxsam** [comment]
@ajbarber 
>Cmake is complaining as the static versions of these libraries (i.e suffix .a) are either not installed on your system (likely) or it can't find them (unlikely)

How do I verify this, and get the right library files? The Docker is a non-option for me, I need to build this specifically for Amazon Linux. (I pretty much have no idea what I'm doing)

**Comment by ffxsam** [comment]
I got stuck installing Boost. Extremely confusing - no matter what I try, `cmake` yells that it can't find the Boost libraries. I'll stick with using the binary + separate library files, as messy as it is. If someone manages to get this thing compiled on an Amazon Linux EC2, let me know.

**Comment by chrisn** [comment]
@ffxsam The `static-link` branch was merged to `master` back in July last year.

**Comment by chrisn** [comment]
IIRC not all of the dependencies are provided with static libraries, some are only distributed with shared libriries (.so). For those, your only option would be to build the library from source.

**Comment by ffxsam** [comment]
Yep, I got that part (had to manually build a bunch of stuff). I've opened a new issue as it's somewhat unrelated at this point (libgd is the last one that won't seem to be built into the executable). #86

**Comment by raedatoui** [comment]
I got this working without static compiling
https://gist.github.com/raedatoui/a0b06c1b239a7ca35b9df29245a82a6f

**Comment by chrisn** [comment]
@raedatoui Awesome! Do you mind if I include this info in our documentation?

**Comment by ffxsam** [comment]
@raedatoui I'm a Docker dummy - I've built the image, but I'm not sure how to find the compiled audiowaveform binary and get it out of the Docker container.

```
$ docker build -t audiowaveform:latest .
Sending build context to Docker daemon  3.584kB
Step 1/5 : FROM amazonlinux:latest
latest: Pulling from library/amazonlinux
[...]
Successfully built 7fc9edb814c7
Successfully tagged audiowaveform:latest
$ docker run --rm audiowaveform:latest ls -CF
bin@   compile.sh*  etc/   lib@    local/  mnt/  proc/	run/   srv/  tmp/  var/
boot/  dev/	    home/  lib64@  media/  opt/  root/	sbin@  sys/  usr/
```

**Comment by ffxsam** [comment]
Giving this a spin. Needs a bit more documentation, like how to actually compile and get the binary out in a single step. Also, if it's not using static compiling, doesn't that mean it relies on external libraries which we don't want in a Lambda environment?

Also, compile.sh needs `cd /` before this line: `git clone https://github.com/bbc/audiowaveform.git`

**Comment by ffxsam** [comment]
I hate to report that this doesn't work. I fired up an EC2 instance using the exact environment Lambda uses (amzn2-ami-hvm-2.0.20190313-x86_64-gp2), and it won't run:

```
[ec2-user@ip-10-0-138-190 ~]$ ./audiowaveform
./audiowaveform: error while loading shared libraries: libsndfile.so.1: cannot open shared object file: No such file or directory
```

**Comment by ffxsam** [comment]
@raedatoui Thanks for the Docker setup. It does compile, but as it's not a static build, it won't work in a Lambda function.

**Comment by ajbarber** [comment]
I've been running the Dockerfile above in a production lambda for about 9 months now.  You build it in the ubuntu container, copy it out as part of an npm build step into a bin folder, then deploy your lambda. Not sure why you would do it any other way for lambda :smile: 

Here is my scripts in package.json if it helps anyone:

`  "scripts": {`
`    "test": "jasmine",`
`    "make-bin": "docker build . -t wavecontainer",`
`    "make-install": "id=$(docker create wavecontainer); docker cp $id:/audiowaveform/build/audiowaveform ./bin/audiowaveform; docker rm -v $id && echo \"Binary available in bin directory\" && exit 0",`
`    "make-clean": "docker rm -f /wavecontainer"`
`},`

**Comment by ffxsam** [comment]
> You build it in the ubuntu container, copy it out as part of an npm build step into a bin folder, then deploy your lambda.

But a binary built on Ubuntu wouldn't run on Lambda since it's Amazon Linux (which I think is based on CentOS).

EDIT: Wait, whaaaaat?! I just tried it and it worked. Since when is this possible? I remember trying to run Ubuntu binaries on Lambda before and it didn't work.

That said, when I compiled audiowaveform, I got it down to 1.6MB. Any idea why yours is 4.7MB?

**Comment by ajbarber** [comment]
It shouldn't matter as it is statically linked, the only thing it is looking for on the lambda instance is a few core c libraries, which as long as they are in sensible places, the runtime linking will find them. So you could probably build on a number of 64 bit linux containers and it would probably work.

To see what is driving the size difference, you would have to build the binary in my container and compare to yours using `ldd audiowaveform`.

**Comment by raedatoui** [comment]
@ffxsam if I add this binary and libmad.so to the lambda package, and add a python wrapper that uses the subprocess module, it wouldnt work?

**Comment by ffxsam** [comment]
@ajbarber Ok, so this is the smaller version:
```
	linux-vdso.so.1 =>  (0x00007ffd8feec000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f7f4cf55000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f7f4cc53000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f7f4ca3d000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f7f4c670000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7f4d259000)
```
and the 4.7MB build:
```
	linux-vdso.so.1 =>  (0x00007ffdb273f000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f588dc3d000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f588d93b000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f588d725000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f588d358000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f588df41000)
```

@raedatoui It might..? Not totally sure.
```

**Comment by ajbarber** [comment]
I see, I thought it might be that the smaller build was not compiling in some library or other, however your screen dump there shows they are the same.

In that case one of the dev dependencies pulled in from the ubuntu package manager could be bloated for some reason, perhaps with debug symbols for example, or some other nested ancillary dependency which was not captured in your alternative smaller build, and which we clearly don't need in a production build.

**Comment by ffxsam** [comment]
Hey, I'm working on a modification of your Dockerfile and script that will build using Amazon Linux. I need help though.. been at this for hours and I'm stuck now.

Unfortunately, on Amazon Linux, the libraries in the yum repository are not static.. (most of the ones needed, anyway). So I had to grab and build those manually. I get to a point and then it bombs out because apparently the `boost148-static` package has every Boost static lib _except_ the regex one. 😕

If there's a better way to do this, someone please stop me so I don't waste more time. 😅 

## Dockerfile
```
FROM amazonlinux:latest
RUN yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
RUN yum -y update
RUN yum -y install wget libstdc++ glibc-devel.x86_64 git make cmake xz \
  zlib-static gcc gcc-c++ libmad-devel libpng-static libtool autogen \
  libid3tag-devel gd-devel boost148-static

COPY ./build.sh build.sh
RUN chmod +x build.sh
RUN ./build.sh
```

## build.sh
```bash
#!/bin/bash

BOOST_INCLUDEDIR=/usr/include/boost148
BOOST_LIBRARYDIR=/usr/lib64/boost148

# Build FLAC libs
wget https://ftp.osuosl.org/pub/xiph/releases/flac/flac-1.3.3.tar.xz
unxz flac-1.3.3.tar.xz
tar xf flac-1.3.3.tar
cd flac-1.3.3
./configure
make install
cp src/libFLAC/.libs/libFLAC-static.a /usr/lib64/libFLAC.a
cd /

# Build Vorbis libs
wget http://downloads.xiph.org/releases/vorbis/libvorbis-1.3.6.tar.gz
tar xf libvorbis-1.3.6.tar.gz
cd libvorbis-1.3.6
./configure
make install
cd /

# Build OGG libs
wget http://downloads.xiph.org/releases/ogg/libogg-1.3.4.tar.gz
tar xf libogg-1.3.4.tar.gz
cd libogg-1.3.4
./configure
make install
cd /

# Build libsndfile
#git clone https://github.com/erikd/libsndfile.git
#cd libsndfile
#sed -i 's/1.14/1.13.1/' configure.ac
#./autogen.sh
#./configure
#make
#make check
#cp src/.libs/libsndfile.* /usr/lib64
#cd /

# Build libmad
wget https://sourceforge.net/projects/mad/files/libmad/0.15.1b/libmad-0.15.1b.tar.gz/download -O libmad-0.15.1b.tar.gz
tar xf libmad-0.15.1b.tar.gz
cd libmad-0.15.1b
sed -i 's/ -fforce-mem//' configure
./configure
make
cp .libs/libmad.a /usr/lib64
cd /

git clone https://github.com/bbc/audiowaveform.git
cd audiowaveform
wget https://github.com/google/googletest/archive/release-1.8.0.tar.gz
tar xzf release-1.8.0.tar.gz
ln -s googletest-release-1.8.0/googletest googletest
ln -s googletest-release-1.8.0/googlemock googlemock
mkdir build
cd build
cmake -D ENABLE_TESTS=0 -D BUILD_STATIC=1 ..
make
strip audiowaveform
```
```

**Comment by ajbarber** [comment]
This package appears to have the regex .a :

https://centos.pkgs.org/8/centos-powertools-x86_64/boost-static-1.66.0-6.el8.x86_64.rpm.html

**Comment by ffxsam** [comment]
I think I ran into a problem trying to install `boost166-static` but I'll try again and see what happens.

**Comment by ffxsam** [comment]
Bah. It needs python 3.6 libs, and unfortunately I don't think I can easily install those via yum.

```
Error: Package: boost169-numpy3-1.69.0-2.el7.x86_64 (epel)
           Requires: libpython3.6m.so.1.0()(64bit)
Error: Package: python36-numpy-1.12.1-3.el7.x86_64 (epel)
           Requires: libpython3.6m.so.1.0()(64bit)
Error: Package: boost169-python3-1.69.0-2.el7.x86_64 (epel)
           Requires: libpython3.6m.so.1.0()(64bit)
Error: Package: python36-numpy-1.12.1-3.el7.x86_64 (epel)
           Requires: python(abi) = 3.6
           Installed: python-2.7.16-3.amzn2.0.1.x86_64 (installed)
               python(abi) = 2.7
               python(abi) = 2.7
           Available: python-2.7.5-58.amzn2.x86_64 (amzn2-core)
               python(abi) = 2.7
               python(abi) = 2.7
           Available: python-2.7.14-58.amzn2.0.2.x86_64 (amzn2-core)
               python(abi) = 2.7
               python(abi) = 2.7
   (etc)
```

**Comment by ffxsam** [comment]
Ok, I've got all the requirements built (manually in most cases). From within the audiowaveform build dir, `cmake -D ENABLE_TESTS=0 -D BUILD_STATIC=1 ..` actually works without issue. But then `make` fails with an error that's way over my head:

```
Linking CXX executable audiowaveform
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `hasOptionValue(boost::program_options::variables_map const&, char const*)':
Options.cpp:(.text+0xf7): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `Options::Options()':
Options.cpp:(.text+0x1b5): undefined reference to `boost::program_options::options_description::options_description(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `Options::parseCommandLine(int, char const* const*)':
Options.cpp:(.text+0xecb): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
Options.cpp:(.text+0xf19): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
Options.cpp:(.text+0xf67): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
Options.cpp:(.text+0xfb5): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::validation_error::validation_error(boost::program_options::validation_error::kind_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)':
Options.cpp:(.text._ZN5boost15program_options16validation_errorC2ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i[_ZN5boost15program_options16validation_errorC5ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i]+0x25): undefined reference to `boost::program_options::validation_error::get_template[abi:cxx11](boost::program_options::validation_error::kind_t)'
Options.cpp:(.text._ZN5boost15program_options16validation_errorC2ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i[_ZN5boost15program_options16validation_errorC5ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i]+0x3d): undefined reference to `boost::program_options::error_with_option_name::error_with_option_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::typed_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>::xparse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) const':
Options.cpp:(.text._ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE6xparseERNS_3anyERKSt6vectorIS7_SaIS7_EE[_ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE6xparseERNS_3anyERKSt6vectorIS7_SaIS7_EE]+0x17): undefined reference to `boost::program_options::validate(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, int)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > boost::program_options::to_internal<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)':
Options.cpp:(.text._ZN5boost15program_options11to_internalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESt6vectorIS7_SaIS7_EERKS8_IT_SaISB_EE[_ZN5boost15program_options11to_internalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESt6vectorIS7_SaIS7_EERKS8_IT_SaISB_EE]+0xa8): undefined reference to `boost::program_options::to_internal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*)':
Options.cpp:(.text._ZN5boost15program_options25basic_command_line_parserIcEC2EiPKPKc[_ZN5boost15program_options25basic_command_line_parserIcEC5EiPKPKc]+0xe1): undefined reference to `boost::program_options::detail::cmdline::cmdline(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::typed_value<double, char>::name[abi:cxx11]() const':
Options.cpp:(.text._ZNK5boost15program_options11typed_valueIdcE4nameB5cxx11Ev[_ZNK5boost15program_options11typed_valueIdcE4nameB5cxx11Ev]+0x1e): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::typed_value<RGBA, char>::name[abi:cxx11]() const':
Options.cpp:(.text._ZNK5boost15program_options11typed_valueI4RGBAcE4nameB5cxx11Ev[_ZNK5boost15program_options11typed_valueI4RGBAcE4nameB5cxx11Ev]+0x1e): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::typed_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>::name() const':
Options.cpp:(.text._ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE4nameEv[_ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE4nameEv]+0x1e): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `boost::program_options::typed_value<int, char>::name[abi:cxx11]() const':
Options.cpp:(.text._ZNK5boost15program_options11typed_valueIicE4nameB5cxx11Ev[_ZNK5boost15program_options11typed_valueIicE4nameB5cxx11Ev]+0x1e): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `void boost::program_options::validate<double, char>(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, double*, long)':
Options.cpp:(.text._ZN5boost15program_options8validateIdcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l[_ZN5boost15program_options8validateIdcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l]+0x4ed): undefined reference to `boost::program_options::invalid_option_value::invalid_option_value(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `void boost::program_options::validate<RGBA, char>(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, RGBA*, long)':
Options.cpp:(.text._ZN5boost15program_options8validateI4RGBAcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS8_ESaIS8_EEESaISC_EEPT_l[_ZN5boost15program_options8validateI4RGBAcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS8_ESaIS8_EEESaISC_EEPT_l]+0x325): undefined reference to `boost::program_options::invalid_option_value::invalid_option_value(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o: In function `void boost::program_options::validate<int, char>(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, int*, long)':
Options.cpp:(.text._ZN5boost15program_options8validateIicEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l[_ZN5boost15program_options8validateIicEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l]+0x16a): undefined reference to `boost::program_options::invalid_option_value::invalid_option_value(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost16exception_detail19error_info_injectorINS_15program_options20invalid_option_valueEEE[_ZTVN5boost16exception_detail19error_info_injectorINS_15program_options20invalid_option_valueEEE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options20invalid_option_valueEEEEE[_ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options20invalid_option_valueEEEEE]+0x38): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTCN5boost10wrapexceptINS_15program_options20invalid_option_valueEEE0_NS_16exception_detail10clone_implINS4_19error_info_injectorIS2_EEEE[_ZTVN5boost10wrapexceptINS_15program_options20invalid_option_valueEEE]+0x38): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost10wrapexceptINS_15program_options20invalid_option_valueEEE[_ZTVN5boost10wrapexceptINS_15program_options20invalid_option_valueEEE]+0x38): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost16exception_detail19error_info_injectorINS_15program_options16validation_errorEEE[_ZTVN5boost16exception_detail19error_info_injectorINS_15program_options16validation_errorEEE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options16validation_errorEEEEE[_ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options16validation_errorEEEEE]+0x38): more undefined references to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const' follow
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcEE[_ZTVN5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost15program_options11typed_valueIicEE[_ZTVN5boost15program_options11typed_valueIicEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost15program_options11typed_valueIdcEE[_ZTVN5boost15program_options11typed_valueIdcEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/audiowaveform.dir/src/Options.cpp.o:(.rodata._ZTVN5boost15program_options11typed_valueI4RGBAcEE[_ZTVN5boost15program_options11typed_valueI4RGBAcEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/audiowaveform.dir/src/Rgba.cpp.o: In function `__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > boost::re_detail_106900::re_is_set_member<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, char, boost::regex_traits<char, boost::cpp_regex_traits<char> >, unsigned int>(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, boost::re_detail_106900::re_set_long<unsigned int> const*, boost::re_detail_106900::regex_data<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, bool)':
Rgba.cpp:(.text._ZN5boost16re_detail_10690016re_is_set_memberIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEjEET_SH_SH_PKNS0_11re_set_longIT2_EERKNS0_10regex_dataIT0_T1_EEb[_ZN5boost16re_detail_10690016re_is_set_memberIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEjEET_SH_SH_PKNS0_11re_set_longIT2_EERKNS0_10regex_dataIT0_T1_EEb]+0x17e): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform_primary[abi:cxx11](char const*, char const*) const'
Rgba.cpp:(.text._ZN5boost16re_detail_10690016re_is_set_memberIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEjEET_SH_SH_PKNS0_11re_set_longIT2_EERKNS0_10regex_dataIT0_T1_EEb[_ZN5boost16re_detail_10690016re_is_set_memberIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEjEET_SH_SH_PKNS0_11re_set_longIT2_EERKNS0_10regex_dataIT0_T1_EEb]+0x4e7): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform[abi:cxx11](char const*, char const*) const'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::unwind_long_set_repeat(bool)':
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE22unwind_long_set_repeatEb[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE22unwind_long_set_repeatEb]+0x5ba): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform_primary[abi:cxx11](char const*, char const*) const'
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE22unwind_long_set_repeatEb[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE22unwind_long_set_repeatEb]+0x8f5): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform[abi:cxx11](char const*, char const*) const'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_long_set()':
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE14match_long_setEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE14match_long_setEv]+0x1b3): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform_primary[abi:cxx11](char const*, char const*) const'
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE14match_long_setEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE14match_long_setEv]+0x613): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform[abi:cxx11](char const*, char const*) const'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_long_set_repeat()':
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE21match_long_set_repeatEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE21match_long_set_repeatEv]+0x25d): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform_primary[abi:cxx11](char const*, char const*) const'
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE21match_long_set_repeatEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE21match_long_set_repeatEv]+0x6a4): undefined reference to `boost::re_detail_106900::cpp_regex_traits_implementation<char>::transform[abi:cxx11](char const*, char const*) const'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_match()':
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE11match_matchEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE11match_matchEv]+0xa6): undefined reference to `boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::maybe_assign(boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_imp()':
MathUtil.cpp:(.text._ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE9match_impEv[_ZN5boost16re_detail_10690012perl_matcherIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISC_EEENS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEE9match_impEv]+0x31a): undefined reference to `boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::maybe_assign(boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)'
CMakeFiles/audiowaveform.dir/src/MathUtil.cpp.o: In function `bool boost::regex_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, char, boost::regex_traits<char, boost::cpp_regex_traits<char> > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >&, boost::basic_regex<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, boost::regex_constants::_match_flags)':
MathUtil.cpp:(.text._ZN5boost11regex_matchIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISB_EEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEEEbT_SJ_RNS_13match_resultsISJ_T0_EERKNS_11basic_regexIT1_T2_EENS_15regex_constants12_match_flagsE[_ZN5boost11regex_matchIN9__gnu_cxx17__normal_iteratorIPKcNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESaINS_9sub_matchISB_EEEcNS_12regex_traitsIcNS_16cpp_regex_traitsIcEEEEEEbT_SJ_RNS_13match_resultsISJ_T0_EERKNS_11basic_regexIT1_T2_EENS_15regex_constants12_match_flagsE]+0xe0): undefined reference to `boost::re_detail_106900::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::construct_init(boost::basic_regex<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, boost::regex_constants::_match_flags)'
collect2: error: ld returned 1 exit status
make[2]: *** [audiowaveform] Error 1
make[1]: *** [CMakeFiles/audiowaveform.dir/all] Error 2
make: *** [all] Error 2
```

Anyone have any ideas?

**Comment by ffxsam** [comment]
I tried manually building boost 1.48 as well, and it wouldn't even build. I'm giving up for now.

**Comment by chrisn** [comment]
I had a quick look into this, maybe a static build isn't needed after all? I used the [amazonlinux][1] Docker image to compile audiowaveform (including building libmad, libid3tag, and libsndfile from source).

Following the information [here][2], I modified `CMakeLists.txt` to add `aws_lambda_package_target`, which creates a zip file with audiowaveform + all its dependencies.

I know very little about Lambda, so I'm not sure what to do next - e.g., how do I run this? how do I send input to it, and where does the output go?

[1]: https://hub.docker.com/_/amazonlinux
[2]: https://aws.amazon.com/blogs/compute/introducing-the-c-lambda-runtime/

**Comment by ffxsam** [comment]
@chrisn You won't have to mess with Lambda at all. If it works in the amazonlinux:2 Docker container, it'll work on Lambda.

I suppose the static build isn't necessary. I'll mess around with a shared lib build and see if it works. In the past, I was opposed to this, but since the introduction of Lambda layers, it's much easier to have several Lambda functions share a single ZIP file of binaries. As long as audiowaveform's dependencies can exist in the same folder as the binary, this should work.

**Comment by ffxsam** [comment]
Hey Chris, would you mind posting your `CMakeLists.txt` file here? In fact, the general steps you took to get this compiling. I'm not sure which parts of the linked article you followed. I'm having trouble even getting the aws-lambda-cpp built:

```
CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
See also "/aws-lambda-cpp/build/CMakeFiles/CMakeOutput.log".
```

**Comment by chrisn** [comment]
Yes, I'll run through the steps again later today, and hopefully produce a Dockerfile that automates everything.

**Comment by ffxsam** [comment]
Thanks! I tried to do the lifting myself based on the work others started on Ubuntu, so I could open a PR with the Dockerfile & build script. Turned out to be more of a headache than I had anticipated, though!

**Comment by chrisn** [comment]
[Here](https://gist.github.com/chrisn/bd19470be8c2b66852ef0346eab84981) is the Dockerfile I made. The zip file is written to `/root/audiowaveform-1.4.0/build/audiowaveform.zip` inside the container. The next step is to copy the zip file out of the docker image and try it on Lambda. Over to you!

**Comment by ffxsam** [comment]
@chrisn Thank you! I tried it and it worked, and it spit out about 20MB worth of stuff. I'll use what you've done here as a basis for a static build Dockerfile and post it here if I get it working.

**Comment by chrisn** [comment]
A static build will mean compiling most of the dependencies from source, as the `-devel` rpms generally don't include static libraries. Have you tried the zip file on Lambda? I'm curious whether it works or not.

**Comment by ffxsam** [comment]
It does indeed work (your version)! I also managed to get a static build made and it's only 2.6MB. My guess is that the shared lib version is including a bunch of system-level libraries that are already present in the Amazon Linux container. (confirmed, I see libstdc++ in there)

Thanks to everyone on this thread for pitching in! I'm happy to share the final Dockerfile and build script if anyone is interested.

**Comment by chrisn** [comment]
Thanks Sam, the Dockerfile and script would be really helpful. As there's been so much interest here, would you be interested in writing a guide, not just for the build but also how to deploy and run it on Lambda? I could include a link in the documentation, or we could include it here event?

**Comment by raedatoui** [comment]
@ffxsam def interested and happy to some testing

**Comment by ffxsam** [comment]
I don't have time at the moment to write a proper guide on Lambda deployment, but I will when I'm able!

I can share my Dockerfile & build script in a bit. Just found an issue or two I need to fix up first.

**Comment by ffxsam** [comment]
@chrisn @raedatoui Here you go:

[audiowaveform.zip](https://github.com/bbc/audiowaveform/files/3811599/audiowaveform.zip)

Please feel free to double-check this. Instructions:

1. `yarn build` to build the container
2. `yarn install` to fire a container to copy the binary from
3. `yarn clean` to remove the container and image

I'm using the binary within Lambda and it works great! And now it's just 1.6MB!

**Comment by raedatoui** [comment]
@ffxsam confirmed its working on a Lambda function! thank you @chrisn and @ffxsam for your help!

I got it working with Python, here is a snippet if someone else can find this useful

```
import boto3
import subprocess

LAMBDA_TASK_ROOT = os.environ.get('LAMBDA_TASK_ROOT')
# binary is added to the lambda package
BIN_PATH = '{root}/audiowaveform'.format(root=LAMBDA_TASK_ROOT) 

client = boto3.client('s3')

file_signed_url = client.generate_presigned_url(
    ClientMethod='get_object',
    ExpiresIn=3600,
    Params={
        'Bucket': bucket,
        'Key': input_key
    }
)

get_file = subprocess.Popen(
    ['curl', file_signed_url],
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE
)

gen_waveform = subprocess.Popen(
    [
        config.BIN_PATH,
        '--input-format', input_format,
        '--output-format', 'dat',
        '-b', '8', '-z', '256'
    ],
    stdin=get_file.stdout,
    stdout=subprocess.PIPE,
)

get_file.stdout.close()

output, err = gen_waveform.communicate()

client.put_object(Bucket=bucket, Key=output_key, Body=output, ContentType='binary/octet')
```

There is probably a better way than using curl and signed urls but this works just fine.

**Comment by TrevorHinesley** [comment]
@ffxsam have you had issues with this throwing an error like the following with flac files:

```
Failed to read file: filename.flac 
File contains data in an unimplemented format.
```

The static library on lambda won't seem to process flac files. This is usually do to libsndfile missing I believe, but from your script I see a section with libsndfile in there. Possible it's not getting pushed into the static build properly?

**Comment by chrisn** [comment]
That error comes from libsndfile (see [here](https://github.com/erikd/libsndfile/blob/326e4533f38cf331fa6b3a6ff4af32ae100c7d27/src/sndfile.c#L81)). It looks like libsndfile has been built without FLAC support enabled.

**Comment by TrevorHinesley** [comment]
Hmm any idea how to configure that? `libsndfile` gets made before `flac` in the above build scripts, but I'm not sure if that affects it. Tried rearranging the build sections to have `libsndfile` build after `flac` and right before `boost` but it threw errors when I went to use it after building it like that:

```
RubyError: StandardError: /opt/audiowaveform: error while loading shared libraries: libmvec.so.1: cannot open shared object file: No such file or directory
```
 
This is definitely not my area of expertise, so any help is much appreciated :)

**Comment by TrevorHinesley** [comment]
Hmm does this just mean I need to statically link a different version of glibc, or perhaps libmvec itself?

**Comment by TrevorHinesley** [comment]
So the ruby 2.5 runtime (which is `amazonlinux:1` in Docker) on Lambda uses an older version of glibc than the `amazonlinux:latest` docker image (aka `amazonlinux:2`). I'm having some other issues deploying using the ruby 2.7 runtime, so I'm trying to adjust the Dockerfile to just use `amazonlinux:1` but it's failing.

**Comment by chrisn** [comment]
If you can share the Dockerfile you're using, I'll take a look.

**Comment by TrevorHinesley** [comment]
Thanks so much! 

# Dockerfiles
Original (from @ffxsam):

```
FROM amazonlinux:latest
RUN yum -y update
RUN yum -y install make cmake3 autogen automake libtool gcc gcc-c++ wget tar \
  gzip zip libcurl-devel zlib-static libpng-static xz git python-devel \
  bzip2-devel which gd-devel

COPY ./build.sh build.sh
RUN chmod +x build.sh
RUN ./build.sh
```

Updated:

```
FROM amazonlinux:1
RUN yum -y update
RUN yum -y install make wget tar gcc gcc-c++
RUN wget https://cmake.org/files/v3.7/cmake-3.7.2.tar.gz && \ 
  tar -xzvf cmake-3.7.2.tar.gz && cd cmake-3.7.2 && \
  ./bootstrap && make && make install
RUN yum -y install cmake autogen automake libtool \
  gzip zip libcurl-devel zlib-static libpng-static xz git python-devel \
  bzip2-devel which gd-devel

COPY ./build.sh build.sh
RUN chmod +x build.sh
RUN ./build.sh
```

Explanation: switched `amazonlinux:latest` (which is now `amazonlinux:2`) to `amazonlinux:1` since that's what the ruby 2.5 runtime uses on Lambda. Also, `cmake3` isn't working in `amazonlinux:1` so I'm trying to build a different version of cmake and use just `cmake` as the command in `build.sh` (see the diffs below, not sure if it will work yet... not my specialty lol).

# build.sh

Original (from @ffxsam):

```
#!/bin/sh

set -ex

# Build libid3tag
wget https://netix.dl.sourceforge.net/project/mad/libid3tag/0.15.1b/libid3tag-0.15.1b.tar.gz
tar xzf libid3tag-0.15.1b.tar.gz
cd libid3tag-0.15.1b
sed -i 's/ -fforce-mem//' configure
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libmad
wget https://netix.dl.sourceforge.net/project/mad/libmad/0.15.1b/libmad-0.15.1b.tar.gz
tar xzf libmad-0.15.1b.tar.gz
cd libmad-0.15.1b
sed -i 's/ -fforce-mem//' configure
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libFLAC
wget https://ftp.osuosl.org/pub/xiph/releases/flac/flac-1.3.0.tar.xz
tar xf flac-1.3.0.tar.xz
cd flac-1.3.0
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libogg
wget http://downloads.xiph.org/releases/ogg/libogg-1.3.4.tar.gz
tar xf libogg-1.3.4.tar.gz
cd libogg-1.3.4
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libvorbis
wget http://downloads.xiph.org/releases/vorbis/libvorbis-1.3.6.tar.gz
tar xf libvorbis-1.3.6.tar.gz
cd libvorbis-1.3.6
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libgd
git clone https://github.com/libgd/libgd.git
cd libgd
mkdir build
cd $_
cmake3 -DBUILD_STATIC_LIBS=1 -DENABLE_PNG=1 ..
make
mv Bin/libgd.a /lib64
cd /

# Build libsndfile (Amazon repo only has earlier 1.0.25 release)
# Requires autogen, automake, and libtool packages
wget https://github.com/erikd/libsndfile/archive/1.0.28.tar.gz
tar xzf 1.0.28.tar.gz
cd libsndfile-1.0.28
sed -i 's/flac >= 1.3.1/flac >= 1.3.0/' configure.ac
./autogen.sh
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build boost
wget https://sourceforge.net/projects/boost/files/boost/1.69.0/boost_1_69_0.tar.gz
tar xf boost_1_69_0.tar.gz
cd boost_1_69_0
./bootstrap.sh --libdir=/lib64 --includedir=/usr/include
./b2 link=static install
cd /

# Build audiowaveform
wget https://github.com/bbc/audiowaveform/archive/1.4.0.tar.gz
tar xzf 1.4.0.tar.gz
mv audiowaveform-1.4.0 audiowaveform
cd $_
mkdir build
cd $_
cmake3 -D ENABLE_TESTS=0 -D BUILD_STATIC=1 ..
make
strip audiowaveform

```

Updated:

```
#!/bin/sh

set -ex

# Build libid3tag
wget https://netix.dl.sourceforge.net/project/mad/libid3tag/0.15.1b/libid3tag-0.15.1b.tar.gz
tar xzf libid3tag-0.15.1b.tar.gz
cd libid3tag-0.15.1b
sed -i 's/ -fforce-mem//' configure
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libmad
wget https://netix.dl.sourceforge.net/project/mad/libmad/0.15.1b/libmad-0.15.1b.tar.gz
tar xzf libmad-0.15.1b.tar.gz
cd libmad-0.15.1b
sed -i 's/ -fforce-mem//' configure
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libogg
wget http://downloads.xiph.org/releases/ogg/libogg-1.3.4.tar.gz
tar xf libogg-1.3.4.tar.gz
cd libogg-1.3.4
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libvorbis
wget http://downloads.xiph.org/releases/vorbis/libvorbis-1.3.6.tar.gz
tar xf libvorbis-1.3.6.tar.gz
cd libvorbis-1.3.6
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libFLAC
wget https://ftp.osuosl.org/pub/xiph/releases/flac/flac-1.3.0.tar.xz
tar xf flac-1.3.0.tar.xz
cd flac-1.3.0
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build libgd
git clone https://github.com/libgd/libgd.git
cd libgd
mkdir build
cd $_
cmake -DBUILD_STATIC_LIBS=1 -DENABLE_PNG=1 ..
make
mv Bin/libgd.a /lib64
cd /

# Build libsndfile (Amazon repo only has earlier 1.0.25 release)
# Requires autogen, automake, and libtool packages
wget https://github.com/erikd/libsndfile/archive/1.0.28.tar.gz
tar xzf 1.0.28.tar.gz
cd libsndfile-1.0.28
sed -i 's/flac >= 1.3.1/flac >= 1.3.0/' configure.ac
./autogen.sh
./configure --disable-shared --libdir=/lib64
make install
cd /

# Build boost
wget https://sourceforge.net/projects/boost/files/boost/1.69.0/boost_1_69_0.tar.gz
tar xf boost_1_69_0.tar.gz
cd boost_1_69_0
./bootstrap.sh --libdir=/lib64 --includedir=/usr/include
./b2 link=static install
cd /

# Build audiowaveform
wget https://github.com/bbc/audiowaveform/archive/1.4.0.tar.gz
tar xzf 1.4.0.tar.gz
mv audiowaveform-1.4.0 audiowaveform
cd $_
mkdir build
cd $_
cmake -D ENABLE_TESTS=0 -D BUILD_STATIC=1 ..
make
strip audiowaveform

```

Explanation: moved libflac under the other libraries so it would hopefully register them as being installed, and add support for them. Also, `cmake3` isn't working in `amazonlinux:1` so I'm trying to build a different version of cmake and use `cmake` (not sure if it will work yet... not my specialty lol).

**Comment by chrisn** [comment]
I got a working build for `amazonlinux:1` using [these files](https://gist.github.com/chrisn/e887d12c543ca6e889041e7aaff76041).

**Comment by TrevorHinesley** [comment]
@chrisn man, I can't think you enough. Unfortunately, `libsndfile` spits this out during the complilation:

```
. . .
checking for flac >= 1.3.1 ... no
checking for ogg >= 1.1.3 ... no
checking for vorbis >= 1.2.3 ... no
checking for vorbisenc >= 1.2.3 ... no


configure: WARNING: *** One or more of the external libraries (ie libflac, libogg and
configure: WARNING: *** libvorbis) is either missing (possibly only the development
configure: WARNING: *** headers) or is of an unsupported version.
configure: WARNING: ***
configure: WARNING: *** Unfortunately, for ease of maintenance, the external libs
configure: WARNING: *** are an all or nothing affair.
. . .
```

So it's not getting built with support for those three still. I'm going to play around with it and see if I can figure out what's going on.

**Comment by TrevorHinesley** [comment]
Very strange... can't seem to get it to see the packages. The final summary printed out by libsndfile's build step shows the proper directories from what I can tell (but confirms that it's not enabling support for FLAC, Ogg or Vorbis):

```
-=-=-=-=-=-=-=-=-=-= Configuration Complete =-=-=-=-=-=-=-=-=-=-

  Configuration summary :

    libsndfile version : .................. 1.0.28

    Host CPU : ............................ x86_64
    Host Vendor : ......................... unknown
    Host OS : ............................. linux-gnu

    Experimental code : ................... no
    Using ALSA in example programs : ...... no
    External FLAC/Ogg/Vorbis : ............ no

  Tools :

    Compiler is Clang : ................... no
    Compiler is GCC : ..................... yes
    GCC version : ......................... 4.8.5
    Sanitizer enabled : ................... no
    Stack smash protection : .............. no

  Installation directories :

    Library directory : ................... /lib64
    Program directory : ................... /usr/local/bin
    Pkgconfig directory : ................. /lib64/pkgconfig
    HTML docs directory : ................. /usr/local/share/doc/
```

**Comment by chrisn** [comment]
Ah, sorry I spoke to soon... The problem was that the libflac, libvorbis etc pkg-config files were being installed in `/usr/lib/pkgconfig` and not `/usr/lib64/pkgconfig`. Setting `--prefix=/usr --libdir=/usr/lib64` fixed it. Here's an [updated gist](https://gist.github.com/chrisn/e887d12c543ca6e889041e7aaff76041).

**Comment by TrevorHinesley** [comment]
You. Are. A. Saint. That worked! Here is the full setup for the Ruby 2.5 runtime on Lambda for anyone that needs it, with FLAC, Ogg and Vorbis support. Appreciate the work you and @ffxsam did. Here's the steps for anyone viewing this thread:

1. yarn build to build the container
1. yarn install to fire a container to copy the binary from
1. yarn clean to remove the container and image

The build is already zipped in the bin folder :)
[audiowaveform.zip](https://github.com/bbc/audiowaveform/files/4239427/audiowaveform.zip)

**Comment by cretueusebiu** [comment]
I've put together a repo for building a binary for AWS Lambda (Amazon Linux 2) with Docker https://github.com/flixier/audiowaveform-aws-lambda

**Comment by peterj35** [comment]
> I got this working on Lambda. To save everyone else the heavy lifting, feel free to grab this:
> 
> http://ffxsam.s3.amazonaws.com/public/audiowaveform-lambda.zip
> 
> When you package up your Lambda function, leave the `bin` and `lib` folders intact. Lambda's environment will automatically have `/var/task/lib` in `LD_LIBRARY_PATH` so no extra work there is needed. However, you will have to add the `bin` folder to the path inside your `exports.handler` function:
> 
> ```js
> process.env.PATH =
>     process.env.PATH + ':' + process.env.LAMBDA_TASK_ROOT + '/bin';
> ```
> 
> Then you can use Node.js to execute `audiowaveform`.
> 
> I know it's not an ideal solution, so if anyone ever figures out how to statically link all those library files into a single binary, please reply to this thread!

I tried this by creating a layer with that zip, but getting the error:

```
"Command failed: audiowaveform -i fileName --pixels-per-second 20 -b 8 -o output.json\n audiowaveform: error while loading shared libraries: libX11.so.6: cannot open shared object file: No such file or directory\n",
```

Is the libX11.so.6 binary missing? Is there a way I can install it?

**Comment by ffxsam** [comment]
@peterj35 I have a much better solution (in fact I've just removed that zip file). Can you email me at `c2FtaEByZWVsY3JhZnRlci5jb20=`? (base64 decode that) 😉

**Comment by ffxsam** [comment]
This has been fully tested and I'm ready to post it here:

https://github.com/reelcrafter/audiotool-builders

Hope that helps!

**Comment by cesarvspr** [comment]
Here is an example of dockerfile combining ffmpeg + audiowaveform

https://gist.github.com/cesarvspr/b4653a85e5f5c3eec3717c6ead870a26

**Comment by sockite** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
